### ISSUE TITLE
Guidelines for the use of "alt" text in icons

### ISSUE THREAD

**Issue by gonzalo-bulnes** [freedomofpress/securedrop#5743]
## Description

@KwadroNaut raised the following question in [Gitter][orig]:

> I noticed that many 'alt-texts' for icons and buttons have names like 'edit user icon' and not 'edit user'. I'd thought these words are mainly for accessibility, and 'edit user button' would be more approriate?

Are there existing guidelines on the use of `alt` text for icons? If so, maybe this issue would be more appropriately located in [documentation repo][docs].

If there are currently no explicit guidelines on the use of `alt` text for icons, then I think it is worth identifying the different contexts in which icons are currently given an `alt` text and how they're rendered by accessibility tools (e.g. VoiceOver, Narrator, Orca). The goal being making sure they do contribute to making SecureDrop more accessible.

  [docs]: https://github.com/freedomofpress/securedrop-docs
  [orig]: https://gitter.im/freedomofpress/securedrop?at=600d3dfb8816425540e843c1

## User Research Evidence

Not as such, but I know for a fact that  some common assumptions about how people use modern text-to-speech tools can lead to counter-effective results. (In particular, the reading of the page is not necessarily linear, most tools have Javascript enabled and more text doesn't necessarily mean better usability. I give a few examples below to start the conversation.)

## User Stories

_None for now, but I can write some if useful once we know more about the scope of this conversation._

**Comment by gonzalo-bulnes** [comment]
Cross-posting from [Gitter][orig]:

> In general, I believe it depends on the context.
>
> If the action "edit user" is described nearby (in which case the icon is only additional visual support) then it is probably better to render that cue literally "edit user icon" than to cause a repetition "edit user Edit user" in the readout.
>
> That would be different if the icon was the only element indicating the availability of the action (e.g. a button containing only an icon).
>
>Beyond that, I think it is fair to ask when the information provided by the icon is redundant (some visual cues that help you locate information when glancing over the page may not bring much if you're going through it jumping from link to link or button to button, for example when using VoiceOver , Narrator or Orca).

  [orig]: https://gitter.im/freedomofpress/securedrop?at=600df2952cb18a437c3c53dc

**Comment by gonzalo-bulnes** [comment]
As a side note, I'm sure that clarifying the intended purpose of the icon's `alt` text would also provide valuable context to the people translating those strings.

**Comment by zenmonkeykstop** [comment]
I followed advice here on a recent update: https://www.w3.org/WAI/tutorials/images/tips/ - in the case where an image is more decorative than anything, guidance is to set blank alt text to cause screen readers to just ignore it.

**Comment by rmol** [comment]
Thanks for raising this, @KwadroNaut and @gonzalo-bulnes. We don't have guidance on alt text, and have generally been working through accessibility issues as they come up, usually in the context of other changes. A dedicated effort to at least normalize handling of buttons/images/titles/alt text, reviewed with screenreaders, is long overdue. We have around 50 occurrences of alt text, so cataloging them could probably be done in a few hours and would be a great start toward fixing the inconsistencies and awkwardness, and coming up with those guidelines.

For translators, with the 1.7.0 release I tried to make use of some of Weblate's contextual features, adding an "alt text" label and explanations to the changed strings, but looking at the results there was still confusion. It would indeed have been better if the alt text were always about the action and not the icon -- in the original example, the source code has an `<i>` element with a `title` that contains "Edit user {username}". If we were to eliminate that element and move its title content to the `alt` text of the image in the link, instead of describing it as an "edit user icon", I think it would be clearer for journalists and translators.

**Comment by ninavizz** [comment]
This issue is beautiful... yes. :)

**Comment by ninavizz** [comment]
Leaving alt-tags blank is often considered ableist and thus not desirable. That is part of why I dislike icons for purely decorative use (such as on the _Submit_ button). A resource I prefer for accessibility when on the fence from a w3 consortium recommendation, is here: https://www.a11yproject.com/

**Comment by zenmonkeykstop** [comment]
Apparently said site makes the same recommendation: https://www.a11yproject.com/checklist/#make-sure-that-decorative-images-use-null-alt-empty-attribute-values

**Comment by KwadroNaut** [comment]
Totally makes sense to leave out an alt-text when as @ninavizz explains. The isseu that triggered me, is an icon that comes without any text. ![image](https://user-images.githubusercontent.com/99527/121337131-80acaf00-c91c-11eb-89c0-260204e08b70.png) I'm not sure what the best practices are there?

I appreciate it that these alt-tags are localized too!

Ironically I'm having troubles with the page @zenmonkeykstop linked… (https://www.a11yproject.com/checklist/#check-to-see-that-keyboard-focus-order-matches-the-visual-layout )

**Comment by zenmonkeykstop** [comment]
I think when we have an icon without any text it's non-decorative by definition, and so it should definitely have a non-null and descriptive alt attribute. So any examples of that we should fix.

**Comment by ninavizz** [comment]
> Apparently said site makes the same recommendation: https://www.a11yproject.com/checklist/#make-sure-that-decorative-images-use-null-alt-empty-attribute-values

Dang. That's a total surprise as a recommendation by them, tbh. But the more I think about it, the more it makes sense for usability in the context of a task, vs in the context of an experience to _consume content_ (such as in docs). In the context of a task it'd be distracting, vs in the context of content consumption it'd contribute to the desired experience. 

Thx for that flag, Kev! :D

**Comment by KwadroNaut** [comment]
I'll try again: remove 'icon,' we don't need to be told it's a button or an icon, that's information that comes from the context. https://www.w3.org/WAI/tutorials/images/functional/#image-used-in-a-button

**Comment by cfm** [comment]
As you suggested originally, @gonzalo-bulnes, a first step here could be to link to guidance (e.g., [this W3C decision-tree](https://www.w3.org/WAI/tutorials/images/decision-tree/)) in `securedrop-docs` and in the checklist used in `securedrop`'s pull-request template.  I'd be happy to draft these changes if there's interest.

I'll also note here an approach we discussed briefly last week our prioritization meeting for #5972:  **Require WCAG compliance along with other linting enforced automatically in CI.**  Here's [an overview of how to do so](https://marcysutton.github.io/a11y-and-ci/#/) and the W3C's [list of such tools](https://www.w3.org/WAI/ER/tools/?q=wcag-20-w3c-web-content-accessibility-guidelines-20&q=command-line-tool), of which [Pa11y](https://pa11y.org/) and [AccessLint](https://accesslint.com/) look especially promising.  But using one of these tools, by itself, would only guarantee the *presence* of `alt` attributes, not their correctness and usefulness in context, so guidance is still the necessary first step.

**Comment by gonzalo-bulnes** [comment]
Thanks for the ping @cfm!

I originally asked about guidelines because in my experience accessibility increases with awareness, and I believe that automated testing alone is unlikely to help as much as we'd think at first. (I may be saying something that's well understood already, please bear with me!) On the other side, practical steps to address the clearest or more common issues can be a greatly effective start.


To clarify what I mean: I am not saying that automated testing is not useful, only that it can be misleading in the absence of guidelines, and cannot be a substitute for training.

The `alt` text conversation in this thread is a good illustration of that: catching a missing `alt` attribute is a good thing. However, no automated testing can tell today if your alternative text is _appropriate_ (as in: has a useful value in the context of a screen reader, including being empty), beyond the fact that it is _present_. (As far as I know about the tooling that's available today, but I actively try to keep up-to-date on the topic.)

Happily, user interfaces tend to rely on recurrent patterns, and **a lot of ground can be covered by providing example implementation for the most common UI elements and patterns**. I've seen that work well within a _design system_ with re-usable components, but even in the absence of re-usable components, example implementations for _a button with text_, _a button with text and an icon_, _a button with an icon but no text_..., can go a long way. (Nice to have: scenarios where one or the other is most appropriate!)

That's how I'd approach setting up guidelines: start from the most recurrent examples, and provide references as further reading, like the ones @cfm, @zenmonkeykstop and @KwadroNaut linked earlier.

Accessibility is a broad topic, and providing practical steps, as far as I can tell, is an effective way to help us improve the user interfaces early, and offer opportunities to everyone to engage further with the topic on their own schedule. Obvsiouly, your experiences may vary! 🙂

**Comment by cfm** [comment]
Absolutely, @gonzalo-bulnes.  For what it's worth, I see a very modest goal for automated accessibility testing, which is to prevent future *regressions* as SecureDrop's accessibility improves.  In other words, the #5972 epic would ideally yield:

1. concrete changes following the recommendations of the 2019 Accessibility Lab audit linked there;
2. distillation of those recommendations in the contributors' guidelines; and
3. automated checking and enforcement of those recommendations just like other code-level linting.

Indeed, I agree with you that (3) is only useful *in addition to* (1) and especially (2).  :-)
